src/coreclr/src/jit/morph.cpp

```diff
@@ -10830,6 +10830,176 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
+    if (op1->OperIs(op2->OperGet()) && op1->OperIs(oper) && !gtIsActiveCSE_Candidate(op2) &&
+        op1->AsOp()->gtGetOp2()->IsCnsIntOrI() && op2->AsOp()->gtGetOp2()->IsCnsIntOrI())
+    {
+        // Don't create a byref pointer that may point outside of the ref object.
+        // If a GC happens, the byref won't get updated. This can happen if one
+        // of the int components is negative. It also requires the address generation
+        // be in a fully-interruptible code region.
+        if (!varTypeIsGC(op1->AsOp()->gtGetOp1()->TypeGet()) && !varTypeIsGC(op2->AsOp()->gtGetOp1()->TypeGet()))
+        {
+            GenTreeIntCon* cns1  = op1->AsOp()->gtGetOp2()->AsIntCon();
+            GenTreeIntCon* cns2  = op2->AsOp()->gtGetOp2()->AsIntCon();
+            const ssize_t  icon1 = cns1->IconValue();
+            const ssize_t  icon2 = cns2->IconValue();
+
+            if (oper == GT_ADD)
+            {
+                if (tree->gtOverflow() || op1->gtOverflow() || op2->gtOverflow())
+                {
+                    return tree;
+                }
+                cns1->SetIconValue(icon1 + icon2);
+            }
+            else if (oper == GT_OR)
+            {
+                cns1->SetIconValue(icon1 | icon2);
+            }
+            else if (oper == GT_XOR)
+            {
+                cns1->SetIconValue(icon1 ^ icon2);
+            }
+            else if (oper == GT_AND)
+            {
+                cns1->SetIconValue(icon1 & icon2);
+            }
+            else
+            {
+                noway_assert(!"unexpected operator");
+            }
+#ifdef _TARGET_64BIT_
+            if (cns1->TypeGet() == TYP_INT)
+            {
+                // we need to properly re-sign-extend or truncate after adding two int constants above
+                cns1->TruncateOrSignExtend32();
+            }
+#endif //_TARGET_64BIT_
+
+            tree->gtOp2 = cns1;
+            DEBUG_DESTROY_NODE(cns2);
+
+            op1->AsOp()->gtOp2 = op2->AsOp()->gtOp1;
+            op1->gtFlags |= (op1->AsOp()->gtOp2->gtFlags & GTF_ALL_EFFECT);
+            DEBUG_DESTROY_NODE(op2);
+            op2 = tree->gtOp2;
+        }
+    }
+
+    if (op2->IsCnsIntOrI() && varTypeIsIntegralOrI(tree->TypeGet()))
+    {
+        // Fold "((x <op> icon1) <op> icon2) to (x <op> (icon1 <op> icon2))"
+        if (op1->OperIs(oper) && !gtIsActiveCSE_Candidate(op1) && op1->AsOp()->gtGetOp2()->IsCnsIntOrI() &&
+            (op1->AsOp()->gtGetOp2()->OperGet() == op2->OperGet()) &&
+            !varTypeIsGC(op1->AsOp()->gtGetOp2()->TypeGet()) && !varTypeIsGC(op2->TypeGet()))
+        {
+            GenTreeIntConCommon* cns1  = op1->AsOp()->gtGetOp2()->AsIntConCommon();
+            GenTreeIntConCommon* cns2  = op2->AsIntConCommon();
+            const ssize_t        icon1 = cns1->IconValue();
+            const ssize_t        icon2 = cns2->IconValue();
+
+            if (oper == GT_ADD)
```

Comment:
that code duplication doesn't look cool, but a function with 6 arguments `GenTreeIntConCommon* fgGetCommutativeResult(oper, icon1, icon2, tree, op1, optional op2)` also doesn't. I am not sure here.

Reply:
Neither am I, tried to simplify it a bit. Let me know if I should extract it to a separate method.